state/indexer/sink/psql/eventsink_psql.go

```diff
@@ -0,0 +1,153 @@
+package psqlsink
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+	proto "github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/pubsub/query"
+	"github.com/tendermint/tendermint/state/indexer"
+	"github.com/tendermint/tendermint/types"
+)
+
+var _ indexer.EventSink = (*PSQLEventSink)(nil)
+
+const (
+	TableEventBlock = "block_events"
+	TableEventTx    = "tx_events"
+	TableResultTx   = "tx_results"
+	DriverName      = "postgres"
+)
+
+// PSQLEventSink is an indexer backend providing the tx/block index services.
+type PSQLEventSink struct {
+	store *sql.DB
+}
+
+func NewPSQLEventSink(connStr string) (indexer.EventSink, *sql.DB, error) {
+	db, err := sql.Open(DriverName, connStr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &PSQLEventSink{
+		store: db,
+	}, db, nil
+}
+
+func (es *PSQLEventSink) Type() indexer.EventSinkType {
+	return indexer.PSQL
+}
+
+func (es *PSQLEventSink) IndexBlockEvents(h types.EventDataNewBlockHeader) error {
+
+	sqlStmt := sq.Insert(TableEventBlock).Columns("key", "value", "height", "type").PlaceholderFormat(sq.Dollar)
+
+	// index the reserved block height index
+	sqlStmt = sqlStmt.Values(types.BlockHeightKey, fmt.Sprint(h.Header.Height), h.Header.Height, "")
+
+	// index begin_block events
+	err := indexEvents(&sqlStmt, h.ResultBeginBlock.Events, types.EventTypeBeginBlock, h.Header.Height)
+	if err != nil {
+		return err
+	}
+
+	// index end_block events
+	err = indexEvents(&sqlStmt, h.ResultEndBlock.Events, types.EventTypeEndBlock, h.Header.Height)
+	if err != nil {
+		return err
+	}
+
+	_, err = sqlStmt.RunWith(es.store).Exec()
+	return err
+}
+
+func (es *PSQLEventSink) IndexTxEvents(txr *abci.TxResult) error {
+
+	// index the tx result
+	var txid uint32
+	sqlStmtTxResult := sq.
+		Insert(TableResultTx).
+		Columns("tx_result").
+		PlaceholderFormat(sq.Dollar).
+		RunWith(es.store).
+		Suffix("RETURNING \"tx_result_id\"")
+	txBz, err := proto.Marshal(txr)
+	if err != nil {
+		return err
+	}
+
+	sqlStmtTxResult = sqlStmtTxResult.Values(txBz)
+
+	// execute sqlStmtTxResult db query and retrieve the txid
+	err = sqlStmtTxResult.QueryRow().Scan(&txid)
+	if err != nil {
+		return err
+	}
+
+	// index the reserved height and hash indices
+	hash := fmt.Sprintf("%X", types.Tx(txr.Tx).Hash())
+
+	sqlStmtEvents := sq.
+		Insert(TableEventTx).
+		Columns("key", "value", "height", "hash", "txid").
+		PlaceholderFormat(sq.Dollar)
+	sqlStmtEvents = sqlStmtEvents.Values(types.TxHashKey, hash, txr.Height, hash, txid)
+	sqlStmtEvents = sqlStmtEvents.Values(types.TxHeightKey, fmt.Sprint(txr.Height), txr.Height, hash, txid)
+	fmt.Println(sqlStmtEvents.ToSql())
+
```

Comment:
Did I miss something? We're not actually indexing the events.

Reply:
Those aren't the events. We also need to index the events, i.e. `txResult.Result.Events`.

Reply:
Ah, right! Will add it. Thanks.